### PR TITLE
[WIP] Support cuda_graph memory release/resume in RL

### PR DESCRIPTION
**Pull request by ryang-max**
## Motivation

Cuda graph will also occupy some GPU memory. To better support colocate RL framework, we can also release this, since it has been supported in [torch_memory_saver](https://github.com/fzyzcjy/torch_memory_saver/pull/30/files)

https://github.com/THUDM/slime/issues/276

## Modifications

Use torch_memory_saver region tagged cuda_graph instead of torch.cuda.graph

## TODO
 - [ ] Currently in test, only 0.1GB memory released, but it seems cuda graph takes 2 GB memory.
 - [x] ~~Tp have issue~~ Resolved by temperarily save the custom all-reduce before capture graph
 - [ ] Modify / test spec decoding

## Checklist

- [x] Format your code according to the [Format code with pre-commit](https://docs.sglang.ai/developer_guide/contribution_guide.html#format-code-with-pre-commit).
- [x] Add unit tests according to the [Run and add unit tests](https://docs.sglang.ai/developer_guide/contribution_guide.html#run-and-add-unit-tests).
- [ ] Update documentation according to [Write documentations](https://docs.sglang.ai/developer_guide/contribution_guide.html#write-documentations).
- [ ] Provide accuracy and speed benchmark results according to [Test the accuracy](https://docs.sglang.ai/developer_guide/contribution_guide.html#test-the-accuracy) and [Benchmark the speed](https://docs.sglang.ai/developer_guide/contribution_guide.html#benchmark-the-speed).

### CODE REVIEW

**Review by yushengsu-thu — COMMENTED** [review]
@ryang-max , I just checked the (Tp have issue) TP>1 issue 
In a tensor parallel (tp_size=2) environment, CUDA graph capture uses a custom all-reduce.
The custom all-reduce requires IPC metadata to share memory across GPUs.
Virtual memory allocated by `torch_memory_saver` is incompatible with IPC operations.

The following modification to `cuda_graph_runner.py` (in `_capture_graph` function) can resolve the above issue - (you can take a look of it to see whether it make sense):
```python
    def _capture_graph(self, graph, pool, stream, run_once_fn):
        if self.model_runner.server_args.enable_memory_saver:
            from torch_memory_saver import torch_memory_saver
            
            # For tensor parallel operations, we need to temporarily disable
            # custom all-reduce during graph capture to avoid IPC compatibility issues
            backup_ca_comm = None
            if self.tp_size > 1:
                # Temporarily disable custom all-reduce for tensor parallel
                backup_ca_comm = self.model_runner.tp_group.ca_comm
                self.model_runner.tp_group.ca_comm = None
            
            try:
                ctx = torch_memory_saver.cuda_graph(
                    graph, pool=pool, stream=stream, tag="graph"
                )
                with ctx:
                    out = run_once_fn()
            finally:
                # Restore custom all-reduce after graph capture
                if backup_ca_comm is not None:
                    self.model_runner.tp_group.ca_comm = backup_ca_comm
        else:
            with self.device_module.graph(graph, pool=pool, stream=stream):
                out = run_once_fn()
        return out  
```  
cc @hebiao064

**Comment by ryang-max** [comment]
> @ryang-max , I just checked the Tp>1 issue In a tensor parallel (tp_size=2) environment, CUDA graph capture uses a custom all-reduce. The custom all-reduce requires IPC metadata to share memory across GPUs. Virtual memory allocated by `torch_memory_saver` is incompatible with IPC operations.
> 
> The following modification to `cuda_graph_runner.py` (in `_capture_graph` function) can resolve the above issue - (you can take a look of it to see whether it make sense):
> 
> ```python
>     def _capture_graph(self, graph, pool, stream, run_once_fn):
>         if self.model_runner.server_args.enable_memory_saver:
>             from torch_memory_saver import torch_memory_saver
>             
>             # For tensor parallel operations, we need to temporarily disable
>             # custom all-reduce during graph capture to avoid IPC compatibility issues
>             backup_ca_comm = None
>             if self.tp_size > 1:
>                 # Temporarily disable custom all-reduce for tensor parallel
>                 backup_ca_comm = self.model_runner.tp_group.ca_comm
>                 self.model_runner.tp_group.ca_comm = None
>             
>             try:
>                 ctx = torch_memory_saver.cuda_graph(
>                     graph, pool=pool, stream=stream, tag="graph"
>                 )
>                 with ctx:
>                     out = run_once_fn()
>             finally:
>                 # Restore custom all-reduce after graph capture
>                 if backup_ca_comm is not None:
>                     self.model_runner.tp_group.ca_comm = backup_ca_comm
>         else:
>             with self.device_module.graph(graph, pool=pool, stream=stream):
>                 out = run_once_fn()
>         return out  
> ```
> 
> cc @hebiao064

Looks good to me! 🚀   I'll add this for further test

**Comment by ryang-max** [comment]
Hi Tom @fzyzcjy, wondering if there's plan to release torch_memory_saver `0.0.9` recently? We may need a new version that support cuda graph offload

**Comment by fzyzcjy** [comment]
hmm I am a bit confused, seems I did sth similar -> https://github.com/sgl-project/sglang/pull/7873

**Comment by ryang-max** [comment]
> hmm I am a bit confused, seems I did sth similar -> #7873

Ok 😆 Not noticed this PR before.... maybe we can just continue follow the #7873 PR. @hebiao064 @yushengsu-thu . If there's no concern I will close this PR.

**Comment by fzyzcjy** [comment]
sad to hear work duplication :/ since this one is WIP and that one has passed tests, yes I agree maybe continue that one would be great

**Comment by ryang-max** [comment]
Clost as this feature already be tracked in #7873
